exporter/awsemfexporter/datapoint.go

```diff
@@ -213,85 +219,184 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
-	positiveBucketCounts := positiveBuckets.BucketCounts()
-	bucketBegin = 0
-	bucketEnd = 0
-	for i := 0; i < positiveBucketCounts.Len(); i++ {
-		index := i + int(positiveOffset)
-		if bucketBegin == 0 {
-			bucketBegin = math.Pow(base, float64(index))
-		} else {
-			bucketBegin = bucketEnd
+	if totalBucketLen == 0 {
+		return []dataPoint{{
+			name: dps.metricName,
+			value: &cWMetricHistogram{
+				Values: []float64{},
+				Counts: []float64{},
+				Count:  metric.Count(),
+				Sum:    metric.Sum(),
+				Max:    metric.Max(),
+				Min:    metric.Min(),
+			},
+			labels:      createLabels(metric.Attributes()),
+			timestampMs: unixNanoToMilliseconds(metric.Timestamp()),
+		}}, true
+	}
+
+	for currentBucketIndex < totalBucketLen {
+		// Create a new dataPointSplit with a capacity of up to splitThreshold buckets
+		split := dataPointSplit{
+			cWMetricHistogram: &cWMetricHistogram{
+				Values: []float64{},
+				Counts: []float64{},
+				Max:    metric.Max(),
+				Min:    metric.Min(),
+				Count:  0,
+				Sum:    0,
+			},
+			length:   0,
+			capacity: splitThreshold,
 		}
-		bucketEnd = math.Pow(base, float64(index+1))
-		metricVal := (bucketBegin + bucketEnd) / 2
-		count := positiveBucketCounts.At(i)
-		if count > 0 {
-			arrayValues = append(arrayValues, metricVal)
-			arrayCounts = append(arrayCounts, float64(count))
+
+		// Only assign `Sum` if this is the first split to make sure the total sum of the datapoints after aggregation is correct.
+		if currentBucketIndex == 0 {
+			split.cWMetricHistogram.Sum = metric.Sum()
 		}
-	}
 
-	// Set count of zero bucket in values/counts array.
-	if metric.ZeroCount() > 0 {
-		arrayValues = append(arrayValues, 0)
-		arrayCounts = append(arrayCounts, float64(metric.ZeroCount()))
-	}
+		if totalBucketLen-currentBucketIndex < splitThreshold {
+			split.capacity = totalBucketLen - currentBucketIndex
+		}
+
+		// Set mid-point of positive buckets in values/counts array.
+		positiveBuckets := metric.Positive()
+		positiveOffset := positiveBuckets.Offset()
+		positiveBucketCounts := positiveBuckets.BucketCounts()
+		bucketBegin = 0
+		bucketEnd = 0
+
+		for split.length < split.capacity && currentPositiveIndex >= 0 {
+			index := currentPositiveIndex + int(positiveOffset)
+			if bucketEnd == 0 {
+				bucketEnd = math.Pow(base, float64(index+1))
+			} else {
+				bucketEnd = bucketBegin
+			}
+			bucketBegin = math.Pow(base, float64(index))
+			metricVal := (bucketBegin + bucketEnd) / 2
+			count := positiveBucketCounts.At(currentPositiveIndex)
+			if count > 0 {
+				split.cWMetricHistogram.Values = append(split.cWMetricHistogram.Values, metricVal)
+				split.cWMetricHistogram.Counts = append(split.cWMetricHistogram.Counts, float64(count))
+				split.length++
+				split.cWMetricHistogram.Count += count
+				if split.length == 1 && currentBucketIndex != 0 {
+					if bucketBegin < bucketEnd {
+						split.cWMetricHistogram.Max = bucketEnd
+					} else {
+						split.cWMetricHistogram.Max = bucketBegin
```

Comment:
Sorry that I'm struggling to understand the logics here. Under what condition will this be hit?

Reply:
Please check the explanation above.